src/metrics/function_error.py

```diff
@@ -0,0 +1,195 @@
+        sequence_length = self.function_class.sequence_length
+        batch_size = self.function_class.batch_size
+        y_dim = self.function_class.y_dim
+        num_models = len(list(models))
+
+        errs = torch.zeros((num_models, num_batches, batch_size, sequence_length, y_dim))
+
+        for i in range(num_batches):
+            params = self.function_class.p_dist.sample()
+            x_batch = self.function_class.x_dist.sample()
+            n = x_batch.shape[2]
+            
+            A = torch.randn(batch_size, n, n)
+            Q, _ = torch.linalg.qr(A, mode="complete")
+            context_space = Q.clone()
+            context_space[:, :, -1] = 0
+            context_space= context_space @ torch.transpose(context_space, dim0=1, dim1=2)
+            test_space = Q.clone()
+            test_space[:, :, :-1] = 0
+            test_space= test_space @torch.transpose(test_space, dim0=1, dim1=2)
+            
+            x_context= x_batch @context_space
+            x_test =   x_batch @test_space
+            
+            for j in range(1, sequence_length):
+                
+                cur_x = x_context.clone()
+                cur_x[:,j] = x_test[:, j]
+
+                if isinstance(params, list):
+                    y_test = self.function_class.evaluate(cur_x, *params)
+                else:
+                    y_test = self.function_class.evaluate(cur_x, params)
+                
+                with torch.no_grad():
+                    errs[:, i, :, j] = torch.stack([
+                        self.metric.evaluate(
+                            y_test,
+                            model.forward(cur_x, y_test)
+                        )
+                        for model in models
+                    ])[:, :, j]
+        
+        errs = torch.reshape(errs, (num_models, num_batches*batch_size, sequence_length))[:, :, 1:]
+
+        return errs
+
+
+class FCErrorSeenPoints(FunctionClassError):
+
+    def evaluate(self, models: Iterable[ContextModel], num_batches: int = 1) -> Iterable[Tensor]:
+        sequence_length = self.function_class.sequence_length
+        batch_size = self.function_class.batch_size
+        y_dim = self.function_class.y_dim
+        num_models = len(list(models))
+        errs = torch.zeros((num_models, num_batches, batch_size, sequence_length, y_dim))
+
+        for i in range(num_batches):
+            params = self.function_class.p_dist.sample()
+            x_batch = self.function_class.x_dist.sample()
+            
+            for j in range(1, sequence_length):
+                x_test = x_batch.clone()
+                perm = torch.stack([torch.randperm(j) for _ in range(batch_size)]).unsqueeze(dim=1) #samples a permutation from 0 to j-1 for each sequence in the batch
+                ind_mat = (perm == 0) + 0.0 #creates a tensor that is one where the zero is in each permutation, and zero everywhere else
```

Comment:
what is the point of generating a permutation here if we just end up caring about where the 0 is? this seems like an overly complicated why of choosing an index uniformly at random

Reply:
we can fix this or understand this better later